### ISSUE TITLE
[SDK] Add transaction generation to the typescript sdk

### ISSUE THREAD

**Issue by davidiw** [aptos-labs/aptos-core#884]
We can submit [signed transactions](https://github.com/aptos-labs/aptos-core/blob/main/types/src/transaction/mod.rs#L479) to our [Rest API](https://github.com/aptos-labs/aptos-core/blob/b9faf4b2eb6d8fae094123306cdc28c05be92a6b/api/src/transactions.rs#L76)

- [x] Understand [bcs](https://github.com/diem/bcs)
- [x] Understand our [hasher](https://github.com/aptos-labs/aptos-core/blob/f05204a649c6f56255ad1168e3153b76a374a084/crates/aptos-crypto/src/hash.rs#L38) and how it applies to [RawTransaction](https://github.com/aptos-labs/aptos-core/blob/b9faf4b2eb6d8fae094123306cdc28c05be92a6b/types/src/transaction/mod.rs#L66)
- [x] Add a guide on how to generate a transaction, what bcs is, and how this works in typescript
- [x] Some vector of transactions generated in rust and their hashes for validating other language implementations -- ideally something we store and verify across all languages.
- [x] Update the [OpenAPI spec](https://github.com/aptos-labs/aptos-core/blob/main/api/doc/openapi.yaml)
- [x] [Typescript SDK](https://github.com/aptos-labs/aptos-core/tree/main/ecosystem/typescript/sdk) support for transaction json blob into transaction bcs blob
- [x] Typescript SDK support for hashing the bcs blob and signing it
- [x] Typescript SDK support for submitting the signed transaction, note, we can submit the json blob + signature to verify

**Comment by jjleng** [comment]
@davidiw: The purpose of hash the payload on the SDK side is to save one hop?

**Comment by davidiw** [comment]
>Update the [OpenAPI spec](https://github.com/aptos-labs/aptos-core/blob/main/api/doc/openapi.yaml)

The goal here is that the current documentation for the spec doesn't make it clear that the RestAPI actually already supports transactions generated completely on the client side.

**Comment by jjleng** [comment]
Let's do some clarifications and be on the same page:
> Update the [OpenAPI spec](https://github.com/aptos-labs/aptos-core/blob/main/api/doc/openapi.yaml)

Here, we want to add more descriptions to OpenAPI spec so that developers understand what each API does? However, OpenAPI is only good for documenting the interfaces and schemas. More descriptions can help developers understand individual interface/schema. To understand how the APIs work together, there should be another doc.

>  Add a guide on how to generate a transaction, what bcs is, and how this works in typescript

How about a guide in https://aptos.dev/?

> [Typescript SDK](https://github.com/aptos-labs/aptos-core/tree/main/ecosystem/typescript/sdk) support for transaction json blob into transaction bcs blob
 Typescript SDK support for hashing the bcs blob and signing it
 Typescript SDK support for submitting the signed transaction, note, we can submit the json blob + signature to verify

Basically, the RestAPI converts client payload to a RawTransaction and then hashes the RawTransaction (BCS RawTransaction first and hashing with SHA3-256). Now we want to do this on the client side with Typescript SDK? Can you please explain why we want to do this on the SDK side? Just to save one hop? If we need to convert and sign RawTransaction on the SDK side, it is better for us to define RawTransaction in OpenAPI spec and codegen the types in Typescript SDK?

**Comment by davidiw** [comment]
> @davidiw: The purpose of hash the payload on the SDK side is to save one hop?

Missed this.

Not only save a hop, which could be upwards of a second saved. It isn't uncommon to prefer that private keys are used to sign only data that can be constructed locally. The current APIs require more trust than some developers will feel comfortable with. On the flip side, they're the right API to enable quick growth in the ecosystem and it's debatable the differences in security... Of course it is our code.

**Comment by davidiw** [comment]
> Let's do some clarifications and be on the same page:
> 
> > Update the [OpenAPI spec](https://github.com/aptos-labs/aptos-core/blob/main/api/doc/openapi.yaml)
> 
> Here, we want to add more descriptions to OpenAPI spec so that developers understand what each API does? However, OpenAPI is only good for documenting the interfaces and schemas. More descriptions can help developers understand individual interface/schema. To understand how the APIs work together, there should be another doc.
> 
The spec file lacks examples for this variety of transaction submission. There's also the need to specify BCS encoding.

> > Add a guide on how to generate a transaction, what bcs is, and how this works in typescript
> 
> How about a guide in https://aptos.dev/?
> 
> > [Typescript SDK](https://github.com/aptos-labs/aptos-core/tree/main/ecosystem/typescript/sdk) support for transaction json blob into transaction bcs blob
> > Typescript SDK support for hashing the bcs blob and signing it
> > Typescript SDK support for submitting the signed transaction, note, we can submit the json blob + signature to verify
> 
> Basically, the RestAPI converts client payload to a RawTransaction and then hashes the RawTransaction (BCS RawTransaction first and hashing with SHA3-256). Now we want to do this on the client side with Typescript SDK? Can you please explain why we want to do this on the SDK side? Just to save one hop? If we need to convert and sign RawTransaction on the SDK side, it is better for us to define RawTransaction in OpenAPI spec and codegen the types in Typescript SDK?

**Comment by jjleng** [comment]
> Some vector of transactions generated in rust and their hashes for validating other language implementations -- ideally something we store and verify across all languages.

Creating some Rust golden tests and store serialized results in golden files should be good, right? We can certainly verify other language implementations based on the golden file content.

**Comment by davidiw** [comment]
> > Some vector of transactions generated in rust and their hashes for validating other language implementations -- ideally something we store and verify across all languages.
> 
> Creating some Rust golden tests and store serialized results in golden files should be good, right? We can certainly verify other language implementations based on the golden file content.

That's a solid starting place. Eventually, we'd want to make an easy clean data set -- maybe golden files can be that data set if constructed well?
